docs/source/reference/database.md

```diff
@@ -1,29 +1,120 @@
+- lookup one user by name
+- list tokens or servers for one user (typically 1-10)
+- etc.
+
+### The database as a limiting factor
+
+As a result of the above transactions in most requests, database performance is the _leading_ factor in JupyterHub's baseline requests-per-second performance, but that cost does not scale significantly with the number of users, active or otherwise.
+However, the database is _rarely_ a limiting factor in JupyterHub performance in a practical sense, because the main thing JupyterHub does is start, stop, and monitor whole servers, which take far more time than any small database transaction, no matter how many records you have or how slow your database is (within reason).
+Additionally, there is usually _very_ little load on the database itself.
+
+By far the most taxing activity on the database is the 'list all users' endpoint, primarily used by the [idle-culling service](https://github.com/jupyterhub/jupyterhub-idle-culler).
+Database-based optimizations have been added to make even these operations feasible for large numbers of users:
+
+1. State filtering on GET /users with [?state=active](./rest-api.html#/default/get_users),
+   which limits the number of results in the query to only the relevant subset (added in JupyterHub 1.3), rather than all users.
+2. [Pagination](api-pagination) of all list endpoints, allowing the request of a large number of resources to be more fairly balanced with other Hub activities across multiple requests (added in 2.0).
+
+:::{note}
+It's important to note when discussing performance and limiting factors and that all of this only applies to requests to `/hub/...`.
+The Hub and its database are not involved in most requests to single-user servers (`/user/...`), which is by design, and largely motivated by the fact that the Hub itself doesn't _need_ to be fast because its operations are infrequent and large.
+:::
+
+## Database backends
+
+JupyterHub supports a variety of database backends via [SQLAlchemy][].
+The default is sqlite, which works great for many cases, but you should be able to use many backends supported by SQLAlchemy.
+Usually, this will mean PostgreSQL or MySQL, both of which are well tested with JupyterHub.
+
+[sqlalchemy]: https://www.sqlalchemy.org
+
+### Default backend: SQLite
+
+The default database backend for JupyterHub is [SQLite](https://sqlite.org).
+We have chosen SQLite as JupyterHub's default because it's simple (the 'database' is a single file) and ubiquitous (it is in the Python standard library).
+It works very well for testing, small deployments, and workshops.
 
 For production systems, SQLite has some disadvantages when used with JupyterHub:
 
-- `upgrade-db` may not work, and you may need to start with a fresh database
+- `upgrade-db` may not always work, and you may need to start with a fresh database
 - `downgrade-db` **will not** work if you want to rollback to an earlier
   version, so backup the `jupyterhub.sqlite` file before upgrading
```

Comment:
Hmmm wait what does downgrade-db do? To me it sounds like upgrade-db / downgrade-db is about upgrading the database schema based on an associated jupyterhub version or similar.

Reply:
It's entirely possible this comment was outdated and should be tested and/or removed. IIRC, weirdly sqlite can _add_ data to the schema, but it couldn't remove it. So adding things almost always works with sqlite (upgrade-db), while removing things almost never did (downgrade-db).